File: Bridge/Resources/.generated/system/guid.js
    Bridge.define("System.Guid", {
        inherits: function () { return [System.IEquatable$1(System.Guid),System.IComparable$1(System.Guid),System.IFormattable]; },
        $kind: "struct",
        statics: {
            error1: "Byte array for GUID must be exactly {0} bytes long",
            valid: null,
            split: null,
            nonFormat: null,
            replace: null,
            rnd: null,
            config: {
                init: function () {
                    this.valid = new RegExp("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", "ig");
                    this.split = new RegExp("^(.{8})(.{4})(.{4})(.{4})(.{12})$");
                    this.nonFormat = new RegExp("^[{(]?([0-9a-f]{8})-?([0-9a-f]{4})-?([0-9a-f]{4})-?([0-9a-f]{4})-?([0-9a-f]{12})[)}]?$", "ig");
                    this.replace = new RegExp("-", "g");
                    this.rnd = new System.Random.ctor();
                    this.empty = System.Guid.empty;
                }
            },
            parse: function (input) {
                return System.Guid.parseExact(input, null);
            },
            parseExact: function (input, format) {
                var r = System.Guid.empty;
                r.parseInternal(input, format, true);
                return r.$clone();
            },
            tryParse: function (input, result) {
                return System.Guid.tryParseExact(input, null, result);
            },
            tryParseExact: function (input, format, result) {
                result.v = System.Guid.empty;
                return result.v.parseInternal(input, format, false);
            },
            newGuid: function () {
                var a = System.Array.init(16, 0);

                System.Guid.rnd.nextBytes(a);

                a[6] = ((a[6] & 15 | 64)) & 255;
                a[8] = ((a[8] & 191 | 128)) & 255;

                return System.Guid.parse(System.Guid.fromBytes(a));
            },
            fromBytes: function (b) {
                if (b == null || b.length !== 16) {
                    throw new System.ArgumentException(System.String.format(System.Guid.error1, 16));
                }

                var s = b.map(System.Guid.makeBinary).join("");

                return System.Guid.split.exec(s).slice(1).join("-");
            },
            makeBinary: function (x) {
                return System.Int32.format((x & 255), "x2");
            },
            getDefaultValue: function () { return new System.Guid(); }
        },
        _a: 0,
        _b: 0,
        _c: 0,
        _d: 0,
        _e: 0,
        _f: 0,
        _g: 0,
        _h: 0,
        _i: 0,
        _j: 0,
        _k: 0,
        config: {
            alias: [
            "format", "System$IFormattable$format"
            ]
        },
        $ctor4: function (uuid) {
            this.$initialize();
            (System.Guid.empty).$clone(this);

            var s = this.parseInternal(uuid, null, true);
        },
        $ctor1: function (b) {
            this.$initialize();
            if (b == null) {
                throw new System.ArgumentNullException("b");
            }

            if (b.length !== 16) {
                throw new System.ArgumentException(System.String.format(System.Guid.error1, 16));
            }

            this._a = (b[3] << 24) | (b[2] << 16) | (b[1] << 8) | b[0];
            this._b = Bridge.Int.sxs((((b[5] << 8) | b[4])) & 65535);
            this._c = Bridge.Int.sxs((((b[7] << 8) | b[6])) & 65535);
            this._d = b[8];
            this._e = b[9];
            this._f = b[10];
            this._g = b[11];
            this._h = b[12];
            this._i = b[13];
            this._j = b[14];
            this._k = b[15];
        },
        $ctor5: function (a, b, c, d, e, f, g, h, i, j, k) {
            this.$initialize();
            this._a = a | 0;
            this._b = Bridge.Int.sxs(b & 65535);
            this._c = Bridge.Int.sxs(c & 65535);
            this._d = d;
            this._e = e;
            this._f = f;
            this._g = g;
            this._h = h;
            this._i = i;
            this._j = j;
            this._k = k;
        },
        $ctor3: function (a, b, c, d) {
            this.$initialize();
            if (d == null) {
                throw new System.ArgumentNullException("d");
            }

            if (d.length !== 8) {
                throw new System.ArgumentException(System.String.format(System.Guid.error1, 8));
            }

            this._a = a;
            this._b = b;
            this._c = c;
            this._d = d[0];
            this._e = d[1];
            this._f = d[2];
            this._g = d[3];
            this._h = d[4];
            this._i = d[5];
            this._j = d[6];
            this._k = d[7];
        },
        $ctor2: function (a, b, c, d, e, f, g, h, i, j, k) {
            this.$initialize();
            this._a = a;
            this._b = b;
            this._c = c;
            this._d = d;
            this._e = e;
            this._f = f;
            this._g = g;
            this._h = h;
            this._i = i;
            this._j = j;
            this._k = k;
        },
        ctor: function () {
            this.$initialize();
        },
        toString: function () {
            return this.format$1(null);
        },
        toString$1: function (format) {
            return this.format$1(format);
        },
        format: function (format, formatProvider) {
            return this.format$1(format);
        },
        toByteArray: function () {
            var g = System.Array.init(16, 0);

            g[0] = (this._a) & 255;
            g[1] = ((this._a >> 8)) & 255;
            g[2] = ((this._a >> 16)) & 255;
            g[3] = ((this._a >> 24)) & 255;
            g[4] = (this._b) & 255;
            g[5] = ((this._b >> 8)) & 255;
            g[6] = (this._c) & 255;
            g[7] = ((this._c >> 8)) & 255;
            g[8] = this._d;
            g[9] = this._e;
            g[10] = this._f;
            g[11] = this._g;
            g[12] = this._h;
            g[13] = this._i;
            g[14] = this._j;
            g[15] = this._k;

            return g;
        },
        parseInternal: function (input, format, check) {
            var r = null;

            if (System.String.isNullOrEmpty(input)) {
                throw new System.ArgumentNullException("input");
            }

            if (System.String.isNullOrEmpty(format)) {
                var m = System.Guid.nonFormat.exec(input);

                if (m != null) {
                    r = m.slice(1).join("-").toLowerCase();
                }
            } else {
                format = format.toUpperCase();

                if (Bridge.referenceEquals(format, "N")) {
                    var m1 = System.Guid.split.exec(input);

                    if (m1 != null) {
                        input = m1.slice(1).join("-");
                    }
                } else if (Bridge.referenceEquals(format, "B") || Bridge.referenceEquals(format, "P")) {
                    var b = Bridge.referenceEquals(format, "B") ? 123 : 40;

                    if ((input.charCodeAt(0) === b) && (input.charCodeAt(((input.length - 1) | 0)) === b)) {
                        input = input.substr(1, ((input.length - 2) | 0));
                    }
                }

                if (input.match(System.Guid.valid) != null) {
                    r = input.toLowerCase();
                }
            }

            if (r != null) {
                this.fromString(r);
                return true;
            }

            if (check) {
                throw new System.FormatException("input is not in a recognized format");
            }

            return false;
        },
        format$1: function (format) {
            var uuid = System.Guid.fromBytes(this.toByteArray());

            switch (format) {
                case "n": 
                case "N": 
                    return uuid.replace(System.Guid.replace, "");
                case "b": 
                case "B": 
                    return System.String.concat(System.String.concat(String.fromCharCode(123), uuid), String.fromCharCode(125));
                case "p": 
                case "P": 
                    return System.String.concat(System.String.concat(String.fromCharCode(40), uuid), String.fromCharCode(41));
                default: 
                    return uuid;
            }
        },
        fromString: function (s) {
            if (System.String.isNullOrEmpty(s)) {
                return;
            }

            s = s.replace(System.Guid.replace, "");

            var r = System.Array.init(16, 0);

            for (var i = 0; i < 16; i = (i + 1) | 0) {
                r[i] = System.Byte.parse(s.substr(((i * 2) | 0), 2), 16);
            }

            this._a = (r[3] << 24) | (r[2] << 16) | (r[1] << 8) | r[0];
            this._b = Bridge.Int.sxs((((r[5] << 8) | r[4])) & 65535);
            this._c = Bridge.Int.sxs((((r[7] << 8) | r[6])) & 65535);
            this._d = r[8];
            this._e = r[9];
            this._f = r[10];
            this._g = r[11];
            this._h = r[12];
            this._i = r[13];
            this._j = r[14];
            this._k = r[15];
        },
        getHashCode: function () {
            var hash = 17;
            hash = hash * 23 + 1684632903;
            hash = hash * 23 + (this._a == null ? 0 : Bridge.getHashCode(this._a));
            hash = hash * 23 + (this._b == null ? 0 : Bridge.getHashCode(this._b));
            hash = hash * 23 + (this._c == null ? 0 : Bridge.getHashCode(this._c));
            hash = hash * 23 + (this._d == null ? 0 : Bridge.getHashCode(this._d));
            hash = hash * 23 + (this._e == null ? 0 : Bridge.getHashCode(this._e));
            hash = hash * 23 + (this._f == null ? 0 : Bridge.getHashCode(this._f));
            hash = hash * 23 + (this._g == null ? 0 : Bridge.getHashCode(this._g));
            hash = hash * 23 + (this._h == null ? 0 : Bridge.getHashCode(this._h));
            hash = hash * 23 + (this._i == null ? 0 : Bridge.getHashCode(this._i));
            hash = hash * 23 + (this._j == null ? 0 : Bridge.getHashCode(this._j));
            hash = hash * 23 + (this._k == null ? 0 : Bridge.getHashCode(this._k));
            return hash;
        },
        equals: function (o) {
            if (!Bridge.is(o, System.Guid)) {
                return false;
            }
            return Bridge.equals(this._a, o._a) && Bridge.equals(this._b, o._b) && Bridge.equals(this._c, o._c) && Bridge.equals(this._d, o._d) && Bridge.equals(this._e, o._e) && Bridge.equals(this._f, o._f) && Bridge.equals(this._g, o._g) && Bridge.equals(this._h, o._h) && Bridge.equals(this._i, o._i) && Bridge.equals(this._j, o._j) && Bridge.equals(this._k, o._k);
        },
        $clone: function (to) {
            var s = to || new System.Guid();
            s._a = this._a;
            s._b = this._b;
            s._c = this._c;
            s._d = this._d;
            s._e = this._e;
            s._f = this._f;
            s._g = this._g;
            s._h = this._h;
            s._i = this._i;
            s._j = this._j;
            s._k = this._k;
            return s;
        }
    });

File: Bridge/System/Guid.cs
using Bridge;

using System.Runtime.CompilerServices;

namespace System
{
    /// <summary>
    /// The Guid data type which is mapped to the string type in Javascript.
    /// </summary>
    [Immutable]
    public struct Guid : IEquatable<Guid>, IComparable<Guid>, IFormattable
    {
        private const string error1 = "Byte array for GUID must be exactly {0} bytes long";

        private static readonly Bridge.Text.RegularExpressions.Regex Valid = new Bridge.Text.RegularExpressions.Regex("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", "i");
        private static readonly Bridge.Text.RegularExpressions.Regex Split = new Bridge.Text.RegularExpressions.Regex("^(.{8})(.{4})(.{4})(.{4})(.{12})$");
        private static readonly Bridge.Text.RegularExpressions.Regex NonFormat = new Bridge.Text.RegularExpressions.Regex("^[{(]?([0-9a-f]{8})-?([0-9a-f]{4})-?([0-9a-f]{4})-?([0-9a-f]{4})-?([0-9a-f]{12})[)}]?$", "i");
        private static readonly Bridge.Text.RegularExpressions.Regex Replace = new Bridge.Text.RegularExpressions.Regex("-", "g");
        private static readonly Random Rnd = new Random();

        public static readonly Guid Empty = new Guid();

        private int _a;
        private short _b;
        private short _c;
        private byte _d;
        private byte _e;
        private byte _f;
        private byte _g;
        private byte _h;
        private byte _i;
        private byte _j;
        private byte _k;

        public Guid(string uuid)
        {
            this = new Guid();

            var s = ParseInternal(uuid, null, true);
        }

        public Guid(byte[] b)
        {
            if (b == null)
            {
                throw new ArgumentNullException("b");
            }

            if (b.Length != 16)
            {
                throw new ArgumentException(string.Format(error1, 16));
            }

            _a = ((int)b[3] << 24) | ((int)b[2] << 16) | ((int)b[1] << 8) | b[0];
            _b = (short)(((int)b[5] << 8) | b[4]);
            _c = (short)(((int)b[7] << 8) | b[6]);
            _d = b[8];
            _e = b[9];
            _f = b[10];
            _g = b[11];
            _h = b[12];
            _i = b[13];
            _j = b[14];
            _k = b[15];
        }

        public Guid(uint a, ushort b, ushort c, byte d, byte e, byte f, byte g, byte h, byte i, byte j, byte k)
        {
            _a = (int)a;
            _b = (short)b;
            _c = (short)c;
            _d = d;
            _e = e;
            _f = f;
            _g = g;
            _h = h;
            _i = i;
            _j = j;
            _k = k;
        }

        public Guid(int a, short b, short c, byte[] d)
        {
            if (d == null)
            {
                throw new ArgumentNullException("d");
            }

            if (d.Length != 8)
            {
                throw new ArgumentException(string.Format(error1, 8));
            }

            _a = a;
            _b = b;
            _c = c;
            _d = d[0];
            _e = d[1];
            _f = d[2];
            _g = d[3];
            _h = d[4];
            _i = d[5];
            _j = d[6];
            _k = d[7];
        }

        public Guid(int a, short b, short c, byte d, byte e, byte f, byte g, byte h, byte i, byte j, byte k)
        {
            _a = a;
            _b = b;
            _c = c;
            _d = d;
            _e = e;
            _f = f;
            _g = g;
            _h = h;
            _i = i;
            _j = j;
            _k = k;
        }

        [Template("Bridge.equalsT({this}, {other})")]
        public extern bool Equals(Guid other);

        [Template("Bridge.compare({this}, {other})")]
        public extern int CompareTo(Guid other);

        public override string ToString()
        {
            return Format(null);
        }

        public string ToString(string format)
        {
            return Format(format);
        }

        public string ToString(string format, IFormatProvider formatProvider)
        {
            return Format(format);
        }

        public byte[] ToByteArray()
        {
            byte[] g = new byte[16];

            g[0] = (byte)(_a);
            g[1] = (byte)(_a >> 8);
            g[2] = (byte)(_a >> 16);
            g[3] = (byte)(_a >> 24);
            g[4] = (byte)(_b);
            g[5] = (byte)(_b >> 8);
            g[6] = (byte)(_c);
            g[7] = (byte)(_c >> 8);
            g[8] = _d;
            g[9] = _e;
            g[10] = _f;
            g[11] = _g;
            g[12] = _h;
            g[13] = _i;
            g[14] = _j;
            g[15] = _k;

            return g;
        }

        public static Guid Parse(string input)
        {
            return ParseExact(input, null);
        }

        public static Guid ParseExact(string input, string format)
        {
            var r = new Guid();
            r.ParseInternal(input, format, true);
            return r;
        }

        public static bool TryParse(string input, out Guid result)
        {
            return TryParseExact(input, null, out result);
        }

        public static bool TryParseExact(string input, string format, out Guid result)
        {
            result = new Guid();
            return result.ParseInternal(input, format, false);
        }

        public static Guid NewGuid()
        {
            var a = new byte[16];

            Rnd.NextBytes(a);

            a[7] = (byte)(a[7] & 0x0f | 0x40);
            a[8] = (byte)(a[8] & 0xbf | 0x80);

            return new Guid(a);
        }

        [Template("{a} === {b}")]
        public static extern bool operator ==(Guid a, Guid b);

        [Template("{a} !== {b}")]
        public static extern bool operator !=(Guid a, Guid b);

        private bool ParseInternal(string input, string format, bool check)
        {
            string r = null;

            if (string.IsNullOrEmpty(input))
            {
                throw new System.ArgumentNullException("input");
            }

            if (string.IsNullOrEmpty(format))
            {
                var m = NonFormat.Exec(input);

                if (m != null)
                {
                    r = m.Slice(1).Join("-").ToLower();
                }
            }
            else
            {
                format = format.ToUpper();

                var p = false;

                if (format == "N")
                {
                    var m = Split.Exec(input);

                    if (m != null)
                    {
                        p = true;
                        input = m.Slice(1).Join("-");
                    }
                }
                else if (format == "B" || format == "P")
                {
                    var b = format == "B" ? new[] { '{', '}' } : new[] { '(', ')' };

                    if ((input[0] == b[0]) && (input[input.Length - 1] == b[1]))
                    {
                        p = true;
                        input = input.Substr(1, input.Length - 2);
                    }
                } else
                {
                    p = true;
                }

                if (p && input.Match(Valid) != null)
                {
                    r = input.ToLower();
                }
            }

            if (r != null)
            {
                FromString(r);
                return true;
            }

            if (check)
            {
                throw new System.FormatException("input is not in a recognized format");
            }

            return false;
        }

        private string Format(string format)
        {
            var s = ((uint)_a).ToString("x8") + ((ushort)_b).ToString("x4") + ((ushort)_c).ToString("x4");
            s = s + (new[] { _d, _e, _f, _g, _h, _i, _j, _k }).Map(MakeBinary).Join("");

            s = Split.Exec(s).Slice(1).Join("-");

            switch (format)
            {
                case "n":
                case "N":
                    return s.Replace(Replace, "");
                case "b":
                case "B":
                    return '{' + s + '}';
                case "p":
                case "P":
                    return '(' + s + ')';
                default:
                    return s;
            }
        }

        private static string MakeBinary(byte x)
        {
            return (x & 0xff).ToString("x2");
        }

        private void FromString(string s)
        {
            if (string.IsNullOrEmpty(s))
            {
                return;
            }

            s = s.Replace(Replace, "");

            var r = new byte[8];

            _a = (int)uint.Parse(s.Substring(0, 8), 16);
            _b = (short)ushort.Parse(s.Substring(8, 4), 16);
            _c = (short)ushort.Parse(s.Substring(12, 4), 16);
            for (var i = 8; i < 16; i++)
            {
                r[i - 8] = byte.Parse(s.Substring(i * 2, 2), 16);
            }

            _d = r[0];
            _e = r[1];
            _f = r[2];
            _g = r[3];
            _h = r[4];
            _i = r[5];
            _j = r[6];
            _k = r[7];
        }
    }
}